akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -882,6 +882,46 @@ trait FlowOps[+Out, +Mat] {
   def recoverWith[T >: Out](pf: PartialFunction[Throwable, Graph[SourceShape[T], NotUsed]]): Repr[T] =
     via(new RecoverWith(-1, pf))
 
+  /**
+   * CompleteOn allows to complete the stream when an upstream error occurs.
+   *
+   * Since the underlying failure signal onError arrives out-of-band, it might jump over existing elements.
+   * This operator can recover the failure signal, but not the skipped elements, which will be dropped.
+   *
+   * '''Emits when''' element is available from the upstream
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes or fails with exception is an instance of the provided type
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def completeOn[T <: Throwable](implicit tag: ClassTag[T]): Repr[Out] = completeOn {
+    case ex if tag.runtimeClass.isInstance(ex) => true
+  }
+
+  /**
+   * CompleteOn allows to complete the stream when an upstream error occurs.
+   *
+   * Since the underlying failure signal onError arrives out-of-band, it might jump over existing elements.
+   * This operator can recover the failure signal, but not the skipped elements, which will be dropped.
+   *
+   * '''Emits when''' element is available from the upstream
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes or fails with exception pf can handle
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def completeOn(pf: PartialFunction[Throwable, Boolean]): Repr[Out] =
+    via(
+      Flow[Out]
+        .recoverWithRetries(-1, pf.andThen((_: Boolean @unchecked) match {
+          case true => Source.empty
+        }))
+        .withAttributes(DefaultAttributes.completeOn and SourceLocation.forLambda(pf)))
```

Comment:
Or we give this one a separate name

Reply:
`completeWhen`?

Reply:
@johanandren How about `completeWhen`?

Reply:
`completeWhen`?

Reply:
I think that's a good name

Reply:
Are we still planning to name it `completeWhen` for the PF variation?

Reply:
@octonato I think so.

Reply:
@johanandren But that way, the scaladsl and javadsl will have different name.

Reply:
The predicate variant in the javadsl can also be `completeWhen`. Why not?

I think it makes lots of sense that a predicated one is suffixed with `When`.